### DISCUSSION TITLE
Code Relocation on NXP RT1060

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#78948] [Q&A] [answered]
Hey, 

I am currently using zephyr_code_relocate() and CONFIG_CODE_DATA_RELOCATION to enhance system performance by relocating certain modules to run from ITCM. I typically load the code in two ways:
1. Using a debugger.
2. Via USB serial, utilizing NXP-MCUBootUtility-3.5.0. For this, I set the board’s switch to “Serial Downloader” mode, load the code (I use the zephyr.s19 file), and upon completion, I disconnect the power, switch to “Internal Boot,” and then reboot the system.

However, code relocation isn’t feasible in the second case due to ITCM being volatile memory. I assume this means I need to use a dedicated bootloader when loading the code via USB serial, whereas it's not necessary when using the debugger.

Is there an example of compiling the code with such a bootloader that loads the required modules into ITCM before the application starts executing?

Thank you for your assistance.

**Comment by ofirshe** [comment]
Hey @DerekSnell,

I recently came across your discussion [#75424](https://github.com/zephyrproject-rtos/zephyr/discussions/75424#discussioncomment-10031573) and found your comments on code relocation for specified project modules.

I’m wondering how to achieve code relocation when loading via USB serial (using NXP-MCUBootUtility) instead of a debugger. You mentioned "configuring the bootloader"—is there any guidance or documentation on how to set this up?

Best regards

**Comment by DerekSnell** [comment, marked as answer]
Hi @ofirshe ,
To summarize the i.MX RT boot flow when any code is located in RAM, the code must be stored in non-volatile memory, like the FlexSPI flash.  The bootloader in ROM will read from the boot memory, configure the SOC and memory, read the boot header stored in that memory to get details about the boot image, and jump to the reset entry function.  The relocation to RAM can be done by the bootloader before the app executes, or the app can relocate code to RAM before the app executes that code.  Regardless of the method that programs the boot image in the memory, including the USB Serial Download mode you are using, this code will need to be relocated after POR.  The debugger gets around this temporarily, because it will also program the code directly to RAM.  But that RAM will be lost when power cycled.

The ROM bootloader does have a RAM Loader feature (and FYI, MCUboot also has a RAM Loader feature).  The boot header for the image includes the RAM load address and the size to relocate.  The bootloader will then copy that portion of the image from the IVT at the start of the image to the RAM load address.  Then it will jump to the reset entry address in the vector table, and the app executes.  There is a Kconfig symbol that makes this easy if `"zephyr,flash"` is in RAM, like ITCM.  The Kconfig was recently updated to [CONFIG_NXP_FLEXSPI_ROM_RAMLOADER](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/common/Kconfig.rom_loader).  In Zephyr v3.7.0 and earlier, it was called [CONFIG_NXP_IMX_RT_ROM_RAMLOADER](https://github.com/zephyrproject-rtos/zephyr/blob/v3.7.0/soc/nxp/imxrt/Kconfig).  

The Zephyr [Code And Data Relocation APIs](https://docs.zephyrproject.org/latest/kernel/code-relocation.html) can also help relocate code to RAM, especially if the whole app does not fit in the RAM, and you want to relocate parts of the app.  With `zephyr_code_relocate()`, the memory is copied after the app starts executing, but early in the startup code.  
> code relocation isn’t feasible in the second case due to ITCM being volatile memory

That is not true, `zephyr_code_relocate()` does enable relocating Code (.text) and rodata to RAM.  Running the [Zperf sample](https://docs.zephyrproject.org/latest/samples/net/zperf/README.html) on the i.MX RT boards uses `zephyr_code_relocate()` to move code to ITCM for performance.  And in the discussion https://github.com/zephyrproject-rtos/zephyr/discussions/75424#discussioncomment-10031573 you referenced, I used `zephyr_code_relocate()` to move more kernel files and libraries to ITCM.  You can reference these as an example.  I am not sure why you feel this does not work without the debugger.  Perhaps you have the `NOCOPY` flag set?

From experience I know that `zephyr_code_relocate()` does not enable relocating everything from flash to RAM.  For relocating the memory that is typically non-volatile, `zephyr_code_relocate()` can relocate the TEXT and RODATA linker sections.  But Zephyr uses other linker sections which go into the `"zephyr,flash"` memory, and `zephyr_code_relocate()` will not relocate those.  Also, some code/rodata in the kernel or SOC files are used in the startup process before this Relocation occurs.  Meaning the Relocated code may need to execute before that code was copied to RAM, which will not work and prevents the app startup from finishing.   When I wrote up the list of files and libraries I relocated in https://github.com/zephyrproject-rtos/zephyr/discussions/75424#discussioncomment-10031573, there are several more kernel files I tried relocating, but found the app failed.  So if you try to relocate some code and the app no longer works, that is likely the reason.  If you need to relocate that code, you will need to do it before the app executes, like with a bootloader.  Or you could use a custom reset vector to do some relocation, and then jump to the Zephyr reset entry.

Best regards

**Reply by ofirshe** [reply]
Hey @DerekSnell,

First thank you for your detailed response. In my case, the entire application doesn't fit in the ITCM, so I’ve tried relocating the major modules (including those you mentioned in [#75424 (comment)](https://github.com/zephyrproject-rtos/zephyr/discussions/75424#discussioncomment-10031573), such as `fsl_enet` and other networking components), which has significantly improved system performance.

I apologize if my previous wording was unclear. I didn’t mean to imply that loading code via USB serial with code relocation enabled wouldn’t work. My question is more about how to build the final application, which will be loaded into flash and includes both the bootloader and the original application to manage the relocation process  of the specified modules after a reset.

As you mentioned, when using a debugger, the bootloader isn’t needed since the ELF file is parsed, and the content is written directly to the appropriate addresses. However, I’d like to understand how to build a "concatenated" application that combines my original app and the bootloader, allowing me to download the zephyr.s19 file to 0x60000000 using the NXP-MCUboot utility. This approach requires a power cycle to switch from serial downloader mode to internal boot mode.

Regarding the issue you mentioned where the application didn’t work after relocating some modules, did you investigate the cause? What could lead to an application that runs from both FlexSPI flash and ITCM failing?  Also, what symptoms did you observe during this failure?

Best regards

**Comment by DerekSnell** [comment]
Hi @ofirshe ,
Since you do not want to relocate all of the app to SRAM, I think your best option is using Zephyr's [Code And Data Relocation APIs](https://docs.zephyrproject.org/latest/kernel/code-relocation.html).  The RAM Load feature in both the ROM bootloader in the SOC and in MCUboot are intended to load the entire image to RAM before the app executes.  But since you are using `zephyr_code_relocate()` already to move specific files to RAM, you do not need to integrate a bootloader with the app for this relocation.  The relocation is managed in the Zephyr startup code.  The build image will store the code in flash, and the app will relocate to RAM.  And this image can be programmed to flash using any method: `west flash`, Serial Downloader mode, a debugger, etc..

For a reference, you can build the [Zperf sample](https://docs.zephyrproject.org/latest/samples/net/zperf/README.html) for the `mimxrt1060_evk` board.  That sample uses `zephyr_code_relocate()` to relocate a few files to ITCM.  You can program it to flash using any method above, and it will boot from flash.  If you still have issues with your app, you can compare with this sample to see what is different.

> Regarding the issue you mentioned where the application didn’t work after relocating some modules, did you investigate the cause? 

Yes, as I shared before, the cause is because the CPU attempts to access the code or RODATA before the relocation function moves it to RAM.  That will not work.

> Also, what symptoms did you observe during this failure?

The app fails in the startup code.  It never gets to `main()`, and the boot banner does not print.  Since the CPU accesses RAM that has not been loaded yet, it usually manifests as a memory access issue, like a hard fault, or an illegal instruction.

Best regards

**Reply by ofirshe** [reply]
> Since you do not want to relocate all of the app to SRAM.

Are you referring to external SDRAM? If so, has anyone attempted to run the entire application from that memory without encountering the issues you mentioned? If you have an example, that would be helpful. Do you think running from SDRAM would improve performance compared to QSPI flash? We're currently seeing very poor networking performance with the default configuration.

> the cause is because the CPU attempts to access the code or RODATA before the relocation function moves it to RAM.

Why is the CPU accessing RAM that hasn't been loaded yet? Shouldn't it only access the RAM after the relocation is complete?

Best Regards

**Comment by DerekSnell** [comment]
Hi @ofirshe 

I was referring to the internal SRAM, since you were asking about ITCM.  But yes, you can execute the code from external SDRAM.  Since SDRAM is quite large, you can use the ROM bootloader to load the entire `zephyr,flash` into SDRAM.  And yes, QSPI latencies are pretty slow, running from SDRAM can be higher performance.  NXP has appnote [AN12437](https://www.nxp.com/docs/en/application-note/AN12437.pdf) that details the different memory options and performance considerations.

Attached is a simple example [SDRAM_hello_world.zip](https://github.com/user-attachments/files/17233102/SDRAM_hello_world.zip) that moves the entire app to SDRAM, and uses the ROM bootloader to load the RAM at boot, before the app executes.  This example was tested on the `mimxrt1060_evkb` board, using the latest main branch with commit e11ddd5.

Some notes about this example: the devicetree overlay splits the SDRAM memory into two regions:
```
	/* Divide SDRAM into two partitions for Code and Data */
	/* Micron MT48LC16M16A2B4-6AIT:G */
	sdram_code: memory@80000000 {
		device_type = "memory";
		reg = <0x80000000 DT_SIZE_M(16)>;
	};
	sdram_data: memory@81000000 {
		device_type = "memory";
		reg = <0x81000000 DT_SIZE_M(16)>;
	};
```
Then the entire app is placed in SDRAM:
```
	chosen {
		zephyr,flash = &sdram_code;
		zephyr,sram = &sdram_data;
	};
```
The board .conf file configures the ROM bootloader to load the image to RAM at boot:
```
CONFIG_NXP_FLEXSPI_ROM_RAMLOADER=y
```
The LMA is adjusted in the app's Kconfig file.  The linker builds the code/rodata for SDRAM at base address 0x8000_0000, but stores in FlexSPI flash at base address 0x1800_0000.  So the LMA adjustment is +0x9800_0000.
```
if NXP_FLEXSPI_ROM_RAMLOADER

# Adjust LMA when zephyr,flash is in SDRAM
FLASH_CHOSEN := zephyr,flash
FLASH_BASE := $(dt_chosen_reg_addr_hex,$(FLASH_CHOSEN))
FLEXSPI_BASE := $(dt_nodelabel_reg_addr_hex,flexspi,1)
config BUILD_OUTPUT_ADJUST_LMA
	default "0x100000000 - $(FLASH_BASE) + $(FLEXSPI_BASE)"

endif # NXP_FLEXSPI_ROM_RAMLOADER
```
Best regards

**Comment by ofirshe** [comment]
Hi @DerekSnell,  

I have two questions related to this topic:  

1. I’ve reached the maximum capacity of the ITCM, but I see there is still available space in the DTCM that could potentially be used for instructions. Do you know how I can utilize this free space in the DTCM for instruction storage?  
   ![Image](https://github.com/user-attachments/assets/059ed2bd-905d-47a0-b75a-9f30e56b2148)

2. Since my image no longer fits entirely in the ITCM, I need to relocate only the critical parts of the code to the TCM regions. As a result, I can’t use the [CONFIG_NXP_FLEXSPI_ROM_RAMLOADER](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/common/Kconfig.rom_loader) flag because it requires relocating the entire image from the FlexSPI boot device into RAM. 
My approach is to create a "package" containing a bootloader and the application binary concatenated together. The bootloader will access a predefined offset where the binary begins, parse the ELF file, and load its contents into RAM based on the ELF-defined sections. This package will reside in NOR flash, and upon reset, the CPU will execute the bootloader, which will then load and execute the application.  
My question is: 
 - Has anyone implemented a similar approach on NXP devices that you could share?  
 - Does Zephyr provide any built-in utilities or examples for creating such a package?  

Thanks in advance!

**Reply by DerekSnell** [reply]
Hi @ofirshe ,

First, in the RT106x devices, the FlexRAM can be configured for an ITCM size greater than 128 KB.  The application note [AN12077](https://www.nxp.com/docs/en/application-note/AN12077.pdf) details how to configure the FlexRAM.  And with Zephyr, the devicetree can configure the FlexRAM, see "Configuring FlexRAM" at [Memory details with Zephyr](https://community.nxp.com/t5/Zephyr-Project-Knowledge-Base/Memory-details-with-Zephyr/ta-p/2008573).

1.  You can execute from DTCM.  Appnote [AN12437](https://www.nxp.com/docs/en/application-note/AN12437.pdf) covers executable memory options supported in the i.MX RT series.  In Zephyr, you can relocate code to DTCM same as relocating code to ITCM, see  [Memory details with Zephyr](https://community.nxp.com/t5/Zephyr-Project-Knowledge-Base/Memory-details-with-Zephyr/ta-p/2008573).

2.  Zephyr's options for partitioning code into multiple memory locations and relocating code are limited, as discussed above.  Some Zephyr users do this partitioning with custom loaders and/or custom linker files.  But I am not aware of a shared example today that demonstrates what you are planning.

Best regards

**Reply by ofirshe** [reply]
Hi @DerekSnell, 

1. Thanks for introducing me to the FlexRAM approach; I think it's an elegant solution that I’ve applied successfully. Specifically, I configured 5 banks for ITCM, 9 banks for DTCM, and 2 banks for OCRAM, as mentioned in [AN12077](https://www.nxp.com/docs/en/application-note/AN12077.pdf) (page 4), where it states that the minimum configuration for OCRAM is 64 KB. Here's my setup:
```
&itcm {
    reg = <0x0 DT_SIZE_K(160)>;
};
&dtcm {
    reg = <0x20000000 DT_SIZE_K(288)>;
};
&ocram {
    reg = <0x20280000 DT_SIZE_K(64)>;
};
&flexram {
    flexram,bank-spec = <FLEXRAM_OCRAM>,
                        <FLEXRAM_OCRAM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_DTCM>,
                        <FLEXRAM_ITCM>,
                        <FLEXRAM_ITCM>,
                        <FLEXRAM_ITCM>,
                        <FLEXRAM_ITCM>,
                        <FLEXRAM_ITCM>;
};
```
As a result:
- I moved all the data to internal SRAM, eliminating the need for the [tweak](https://github.com/zephyrproject-rtos/zephyr/discussions/81162) to locate the interrupt and main stacks in DTCM.
- I successfully relocated large parts of the code to ITCM.
I achieved the data movement by replacing `zephyr,sram = &sdram0;` with `zephyr,sram = &dtcm;`. However, I faced an issue when trying to declare the chosen `zephyr,dtcm` to be `sdram0`. This is necessary for me because once the internal DTCM space is exhausted, I want to utilize the free 32 MB of SDRAM as an overflow memory if needed.
Even if I set the device tree as follows:
```
zephyr,dtcm = &sdram0;
```
The compilation output still indicates that the `zephyr,dtcm` node refers to the same node as RAM (the internal DTCM), which can lead to data access violations. Here's a visual reference from the compilation output:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/705b7b1c-27b1-488f-9bb0-628205a0a5c2" />
When I remove `zephyr,dtcm` entirely from the DTS, I see that zero bytes are used in DTCM. Additionally, the bytes from the previous compilation are now added to the RAM section, resolving the potential data access violation. 
<img width="393" alt="image" src="https://github.com/user-attachments/assets/7c952fef-b886-405b-92d2-c2170829a175" />
However, this prevents me from utilizing the 32 MB SDRAM for any buffer allocations since there’s no section in the final ELF referring to this memory region.
Can you suggest an elegant solution to resolve this issue so I can utilize SDRAM when necessary? For example, I’d like to configure it in such a way that once the internal DTCM space is exhausted, Zephyr automatically utilizes the 32 MB SDRAM as overflow memory. Is it possible, or do I have to manually link my buffer to that SDRAM region? Additionally, if you have any comments or suggestions to improve the FlexRAM setup I implemented above, I’d be happy to hear.

2. I missed your [response](https://github.com/zephyrproject-rtos/zephyr/discussions/78948#discussioncomment-10811897) earlier, where you wrote:

> But since you are using zephyr_code_relocate() already to move specific files to RAM, you do not need to integrate a bootloader with the app for this relocation. The relocation is managed in the Zephyr startup code. The build image will store the code in flash, and the app will relocate to RAM.

I’ve confirmed that you’re correct. When I flash the ELF image using the NXP Serial Downloader Utility, the relocation is handled by Zephyr. To verify this, I added a print statement inside the main function, and it worked as expected. This means I don’t need to integrate a bootloader, saving me a significant amount of work - so thank you again. The relocation process is handled by the `data_copy_xip_relocation()` function in `code_relocation.c.`

I would be very grateful if you could help me with the first question and share how you would approach this problem
Best regards.

**Reply by DerekSnell** [reply]
Hi @ofirshe ,
I am not aware that the GCC linker can use overflow memory like this.  My experience is the linker will give an error if a section overflows and does not fit all the memory placed in that section.  Since you want most of your data to default in DTCM, I would configure `zephyr,sram` to `&dtcm`.  And then move specific data or source files to SDRAM.  For options to place data in other sections, see "Locating data to RAM" at [Memory details with Zephyr](https://community.nxp.com/t5/Zephyr-Project-Knowledge-Base/Memory-details-with-Zephyr/ta-p/2008573).

BTW, external SDRAM is not DTCM.  Tightly Coupled Memory (TCM) is on-chip SRAM that clocks at the same frequency as the CPU, provides the best performance, and is not cached.  By comparison, external SDRAM is slower and is usually cached.  So I would not configer `zephyr,dtcm = &sdram0`.

Best regards

**Reply by ofirshe** [reply]
Hi @DerekSnell,

You are absolutely correct that DTCM offers the best performance, which is why I moved all the data to DTCM by default. However, if I set `zephyr,sram = &dtcm`, I’m unsure how to utilize SDRAM in cases where I need additional memory. My intention here is to ensure I have the flexibility to allocate certain future buffers or data structures to SDRAM when the available DTCM memory is fully utilized. For example, if I needed to define a buffer in SDRAM, I’d want to be able to do something like this:
`static __sdram uint8_t buf[1024];`
But I’m not sure how to achieve this with `zephyr,sram` configured to `&dtcm`.

Best regards, 
Ofir.

**Reply by ofirshe** [reply]
I've resolved the last issue mentioned [here](https://github.com/zephyrproject-rtos/zephyr/discussions/78948#discussioncomment-11719365) by creating a custom linker script:
```
/* Include the base linker file */
#include <linker.ld>

/* Define the SDRAM data section with load address */
SECTIONS {
    .sdram_data : {
        __sdram_data_load_start = LOADADDR(.sdram_data);  /* Load address in FLASH */
        __sdram_data_start = .;  /* Start of section in SDRAM */

        /* Include all data in the sdram_data section */
        *(.sdram_data)
        *(.sdram_data.*)
        /* 
         * Align the section end to _region_min_align to ensure proper alignment.
         * The _region_min_align value is defined in the base linker file.
         */
        . = ALIGN(_region_min_align);
        __sdram_data_end = .;  /* End of the section */
    } > SDRAM AT > FLASH /* Place .sdram_data in SDRAM for execution, but load its initial content from FLASH. */
}
```
I also created an initialization function that relocates data from FLASH to SDRAM utilizing `CONFIG_SOC_PREP_HOOK`:
#include <zephyr/kernel.h>
#include <zephyr/sys/printk.h>

#ifdef CONFIG_SOC_PREP_HOOK
```
extern char __sdram_data_load_start[];
extern char __sdram_data_start[];
extern char __sdram_data_end[];
void z_early_memset(void *dst, int c, size_t n);
void z_early_memcpy(void *dst, const void *src, size_t n);

/*  
 * This function is expected to be called from the `z_prep_c()` function in  
 * `zephyr/arch/arm/core/cortex_m/prep_c.c` when `CONFIG_SOC_PREP_HOOK` is  
 * enabled. The SoC hook is executed after the reset vector.  
 */
void soc_prep_hook(void)
{
    /* Clear and copy sdram_data from the load address to its destination in SDRAM */
    z_early_memset(&__sdram_data_start, 0,
                   (uintptr_t)&__sdram_data_end - (uintptr_t)&__sdram_data_start);

    z_early_memcpy((void *)&__sdram_data_start,
                   &__sdram_data_load_start,
                   __sdram_data_end - __sdram_data_start);
}
#endif
```
